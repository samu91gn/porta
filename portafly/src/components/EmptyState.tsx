import React from 'react'
import {
  Title,
  EmptyState,
  EmptyStateVariant,
  EmptyStateIcon,
  Bullseye,
  EmptyStateBody
} from '@patternfly/react-core'
import { CubesIcon, SearchIcon } from '@patternfly/react-icons'
import { IRow } from '@patternfly/react-table'

export interface ISimpleEmptyStateProps {
  msg: string
}

const SimpleEmptyState: React.FunctionComponent<ISimpleEmptyStateProps> = ({ msg }) => (
  <EmptyState variant={EmptyStateVariant.full}>
    <EmptyStateIcon icon={CubesIcon} />
    <Title headingLevel="h5" size="lg">
      {msg}
    </Title>
  </EmptyState>
)

type ICreateTableEmptyState = (title: string, body: string, button?: JSX.Element) => IRow[]

const CreateTableEmptyState: ICreateTableEmptyState = (title, body, button) => [{
  heightAuto: true,
  cells: [{
    props: { colSpan: 8 },
    title: (
      <Bullseye>
        <EmptyState variant={EmptyStateVariant.small}>
          <EmptyStateIcon icon={SearchIcon} />
          <Title headingLevel="h2" size="lg">
            {title}
          </Title>
          <EmptyStateBody>
            {body}
          </EmptyStateBody>
          {button}
        </EmptyState>
      </Bullseye>
    )
  }]
}]

export { SimpleEmptyState, CreateTableEmptyState }
