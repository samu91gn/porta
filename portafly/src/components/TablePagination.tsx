import React, { FunctionComponent, useMemo, useReducer } from 'react'
import { Pagination, OnSetPage, OnPerPageSelect } from '@patternfly/react-core'

interface IPagination {
  itemCount: number
  state: PaginationState
  dispatch: React.Dispatch<PaginationAction>
}

const perPageOptions = [
  { title: '5', value: 5 },
  { title: '20', value: 20 },
  { title: '50', value: 50 }
]

const TablePagination: FunctionComponent<IPagination> = ({
  itemCount,
  state,
  dispatch
}) => {
  const { perPage, page } = state

  const onSetPage: OnSetPage = (ev, newPage, _perPage, startIdx, endIdx) => {
    dispatch({
      type: 'setPage',
      payload: {
        page: newPage,
        startIdx,
        endIdx
      }
    })
  }

  const onPerPageSelect: OnPerPageSelect = (ev, newPerPage, newPage, startIdx, endIdx) => {
    dispatch({
      type: 'setPerPage',
      payload: {
        page: newPage,
        perPage: newPerPage,
        startIdx,
        endIdx
      }
    })
  }

  const isCompact = useMemo(() => (itemCount / perPage) < 3, [itemCount, perPage])

  return (
    <Pagination
      itemCount={itemCount}
      perPage={perPage}
      page={page}
      onSetPage={onSetPage}
      onPerPageSelect={onPerPageSelect}
      perPageOptions={perPageOptions}
      isCompact={isCompact}
    />
  )
}

type PaginationState = {
  page: number
  perPage: number
  startIdx?: number
  endIdx?: number
}

type PaginationAction = {
  type: 'setPage' | 'setPerPage'
  payload: Partial<PaginationState>
}

const paginationReducer = (state: PaginationState, action: PaginationAction):
PaginationState => {
  switch (action.type) {
    case 'setPage':
    case 'setPerPage':
      return { ...state, ...action.payload }
    default:
      return { ...state }
  }
}

const usePaginationReducer = () => useReducer(paginationReducer, {
  page: 0,
  perPage: perPageOptions[0].value,
  startIdx: 0,
  endIdx: perPageOptions[0].value
})

export { TablePagination, usePaginationReducer }
