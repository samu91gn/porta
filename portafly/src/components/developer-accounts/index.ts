export * from './ActionsDropdown'
export * from './BulkActionSelector'
export * from './SearchWidget'
export * from './DeveloperAccountsTable'
export * from './SendEmailModal'
export * from './ChangePlanModal'
export * from './ChangeStateModal'
