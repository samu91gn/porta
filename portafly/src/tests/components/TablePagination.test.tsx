import React from 'react'

import { render } from 'tests/custom-render'
import { TablePagination } from 'components'

it('should render compacted when there are 1-2 pages only', () => {
  const wrapper = render(<TablePagination itemCount={5} state={{ perPage: 5 }} dispatch={jest.fn()} />)
  expect(wrapper.container.querySelector('.pf-m-compact')).toBeInTheDocument()
})

it('should render NOT compacted when there are 3 or more pages', () => {
  const wrapper = render(<TablePagination itemCount={5} state={{ perPage: 1 }} dispatch={jest.fn()} />)
  expect(wrapper.container.querySelector('.pf-m-compact')).not.toBeInTheDocument()
})
