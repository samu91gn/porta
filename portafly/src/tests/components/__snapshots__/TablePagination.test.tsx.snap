// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render NOT compacted when there are 3 or more pages 1`] = `
<div
  class="pf-c-pagination"
  id="pagination-options-menu"
>
  <div
    class="pf-c-pagination__total-items"
  >
    <b>
      1
       - 
      1
    </b>
     
    of 
    <b>
      5
    </b>
     
    
  </div>
  <div
    class="pf-c-options-menu"
  >
    <div
      class="pf-c-options-menu__toggle pf-m-plain pf-m-text"
    >
      <span
        class="pf-c-options-menu__toggle-text"
      >
        <b>
          1
           - 
          1
        </b>
         
        of 
        <b>
          5
        </b>
         
        
      </span>
      <button
        aria-expanded="false"
        aria-label="Items per page"
        class="  pf-c-options-menu__toggle-button"
        id="pagination-options-menu-toggle"
        type="button"
      >
        <svg
          aria-hidden="true"
          class=""
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            transform=""
          />
        </svg>
      </button>
    </div>
  </div>
  <nav
    aria-label="Pagination"
    class="pf-c-pagination__nav"
  >
    <button
      aria-label="Go to first page"
      class="pf-c-button pf-m-plain"
      data-action="first"
      disabled=""
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 448 512"
        width="1em"
      >
        <path
          d="M223.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L319.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L393.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34zm-192 34l136 136c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9L127.9 256l96.4-96.4c9.4-9.4 9.4-24.6 0-33.9L201.7 103c-9.4-9.4-24.6-9.4-33.9 0l-136 136c-9.5 9.4-9.5 24.6-.1 34z"
          transform=""
        />
      </svg>
    </button>
    <button
      aria-label="Go to previous page"
      class="pf-c-button pf-m-plain"
      data-action="previous"
      disabled=""
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 256 512"
        width="1em"
      >
        <path
          d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
          transform=""
        />
      </svg>
    </button>
    <div
      class="pf-c-pagination__nav-page-select"
    >
      <input
        aria-label="Current page"
        class="pf-c-form-control"
        max="5"
        min="1"
        type="number"
        value="1"
      />
      <span
        aria-hidden="true"
      >
        of 
        5
      </span>
    </div>
    <button
      aria-label="Go to next page"
      class="pf-c-button pf-m-plain"
      data-action="next"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 256 512"
        width="1em"
      >
        <path
          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
          transform=""
        />
      </svg>
    </button>
    <button
      aria-label="Go to last page"
      class="pf-c-button pf-m-plain"
      data-action="last"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 448 512"
        width="1em"
      >
        <path
          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
          transform=""
        />
      </svg>
    </button>
  </nav>
</div>
`;

exports[`should render compacted when there are 1-2 pages only 1`] = `
<div
  class="pf-c-pagination pf-m-compact"
  id="pagination-options-menu"
>
  <div
    class="pf-c-pagination__total-items"
  >
    <b>
      1
       - 
      5
    </b>
     
    of 
    <b>
      5
    </b>
     
    
  </div>
  <div
    class="pf-c-options-menu"
  >
    <div
      class="pf-c-options-menu__toggle pf-m-plain pf-m-text"
    >
      <span
        class="pf-c-options-menu__toggle-text"
      >
        <b>
          1
           - 
          5
        </b>
         
        of 
        <b>
          5
        </b>
         
        
      </span>
      <button
        aria-expanded="false"
        aria-label="Items per page"
        class="  pf-c-options-menu__toggle-button"
        id="pagination-options-menu-toggle"
        type="button"
      >
        <svg
          aria-hidden="true"
          class=""
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            transform=""
          />
        </svg>
      </button>
    </div>
  </div>
  <nav
    aria-label="Pagination"
    class="pf-c-pagination__nav"
  >
    <button
      aria-label="Go to previous page"
      class="pf-c-button pf-m-plain"
      data-action="previous"
      disabled=""
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 256 512"
        width="1em"
      >
        <path
          d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
          transform=""
        />
      </svg>
    </button>
    <button
      aria-label="Go to next page"
      class="pf-c-button pf-m-plain"
      data-action="next"
      disabled=""
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 256 512"
        width="1em"
      >
        <path
          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
          transform=""
        />
      </svg>
    </button>
  </nav>
</div>
`;
